README.md
# Table of Contents

1.  [Prologue](#org8b6bf34)
2.  [Nalude](#org0812938)
3.  [Structure of Functions](#org658c0e2)
    1.  [Folds and Traversals](#orgebe257d)
    2.  [Lists](#org8fd809b)
    3.  [Miscellaneous](#orgd3b3bf1)
    4.  [String](#org8147070)
    5.  [Specials](#org2de95e8)
    6.  [Tuples](#orge6105ca)
    7.  [Zip and Unzip](#orgcdca5f7)
    8.  [Extra](#org5b1da39)
4.  [Get Start](#orgccf98f8)
    1.  [Install](#org1d1b386)
5.  [Epoligue](#orgd300cef)
    1.  [History](#org1d83dae)
        1.  [Version 0.3.0](#orga174a69)
        2.  [Version 0.2.0](#orga9ce38e)
        3.  [Version 0.1.0](#orgad971e1)



<a id="org8b6bf34"></a>

# Prologue

I like Haskell, also hope some of the prelude functions in Haskell can be used in Python.

Wish you enjoy coding with nalude.


<a id="org0812938"></a>

# Nalude

Nalude is a standard module, which is inspired by Haskell's Prelude.  Thus, nalude is also a
library of functional programming.

Nalude doesn't explicitly distinguish between Applicative's or monadic function since they are not
well differentiated in the basic structure of Python.  Of course, the main reason is that I am too
fresh.

The following functions, Nalude tries to implement:

1.  Folds and Traversals
2.  Lists
3.  Miscellaneous
4.  Specials
5.  Strings
6.  Tuples
7.  Zip and Unzip

At the same time, it also implements some functions which are only available in Python and are
included in \`extra\`.


<a id="org658c0e2"></a>

# Structure of Functions


<a id="orgebe257d"></a>

## Folds and Traversals

-   **foldr(f: Callable[[b, a], b], init: b, t: Sequence[a]):** Right-associative fold of a sequence.
-   **foldl(f, init, t):** Left-associative fold of an iterable.  The same as `reduce`.
-   **foldr1(f, t):** A variant of foldr without base case.
-   **foldl1(f, t):** A variant of foldl without base case.
-   **product(nums):** Computes the product of the numbers.
-   **(HELP WANTED) traverse:** Map each element of a structure to an action, evaluate these actions
    from left to right, and collect the results.
-   **(HELP WANTED) sequence:** Evaluate each actions in the structure from left to right, and collect
    the results.


<a id="org8fd809b"></a>

## Lists

-   **head(xs):** Extract the first element of an iterable.
-   **last(xs):** Extract the last element of an iterable.
-   **null(xs):** Test whether the sequence is empty.
-   Infinite lists
    -   **iterate(f, x):** Yield a tuple of repeated applications of f to x.
    -   **repeat(x):** Repeat x.
    -   **replicate(n, x):** Return a list of length n with x the value of every element.
    -   **cycle(xs):** Tie an iterable to infinite circuler one.

    -   Sublists
        -   **tail(xs):** Extract the elements after the head of a list.
        -   **init(xs):** Extract all the elements of an iterable except the last one.
        -   **take(n, xs):** Return the first n elements of sequence xs.
        -   **drop(n, xs):** Return the remaining elements of xs after the first n elements.
        -   **splitat(n, xs):** Return a tuple where the first element is xs prefix of length n and
            second element is the remainder of the sequence xs.
        -   **takewhile(p, xs):** Return the longest prefix of xs of elements that satisfy predicate p.
        -   **dropwhile(p, xs):** Return the suffix remaining after takewhile(p, xs).
        -   **(LAZY ONE HELP WANTED) span(p, xs):** Equal to (takewhile(p, xs), dropwhile(p, xs)).
        -   **(LAZY ONE HELP WANTED) break\_(p, xs):** Equal to (takewhile(not\_(p), xs),
            dropwhile(not\_(p), xs)).


<a id="orgd3b3bf1"></a>

## Miscellaneous

-   **id\_(x):** Identity function.
-   **const(x, \_):** Evaluates to x for all inputs.
-   **o(f1, f2):** (.) Function composition
-   **flip(f):** Flip takes its two arguments into the reverse order of f.
-   **until(p, f, x):** Yield the result of applying f until p holds.


<a id="org8147070"></a>

## String

-   **lines(s):** Break up a string into a list of strings at newline characters.
-   **unlines(xs):** The inverse operation of lines, append a newline to each.
-   **words(s):** Break a string up into a list of words, which were delimited by white space.
-   **unwords(xs):** The inverse operation of words, join words with space.


<a id="org2de95e8"></a>

## Specials

-   **not\_(f):** Boolean "not".
-   **all\_(p, xs):** Determine whether all elements of the structure satisfy the p.
-   **any\_(p, xs):** Determine whether sny elements of the structure satisfies the p.
-   **concat(xss):** The concatenation of all the elements of a container of Sequences.
-   **concatmap(f, xss):** Map a function over all the elements of a container and concatenate the
    resulting lists.


<a id="orge6105ca"></a>

## Tuples

-   **fst(t):** Extract the first component of a tuple.
-   **snd(t):** Extract the second component of a tuple.
-   **curry(f, a, b):** Converts an uncurried function to a curried function.
-   **uncurry(f, ab@(a, b)):** Converts a curried function to a function on pairs.


<a id="orgcdca5f7"></a>

## Zip and Unzip

-   **zipwith(f, \*seqs):** Zipwith is map(f, zip), but f accept separate args instead of tuple
-   **unzip(pairs):** Transform an iterable of pairs into a tuple of sequence. (Not lazy)


<a id="org5b1da39"></a>

## Extra

-   **flatten(xs,\*,ignore=(dict,)):** Flatten iterables of iterable to a single iterable.  It will
    ignore instances in `ignore` tuple.


<a id="orgccf98f8"></a>

# Get Start


<a id="org1d1b386"></a>

## Install

    pip install nalude


<a id="orgd300cef"></a>

# Epoligue


<a id="org1d83dae"></a>

## History


<a id="orga174a69"></a>

### Version 0.3.0

-   **Data:** <span class="timestamp-wrapper"><span class="timestamp">&lt;Sun Mar 10, 2019&gt;</span></span>
-   **Add:** -   Custom ignore when flatten iterable.


<a id="orga9ce38e"></a>

### Version 0.2.0

-   **Data:** <span class="timestamp-wrapper"><span class="timestamp">&lt;Fri Feb 15, 2019&gt;</span></span>
-   **Add:** -   flatten in extra.


<a id="orgad971e1"></a>

### Version 0.1.0

-   **Data:** <span class="timestamp-wrapper"><span class="timestamp">&lt;Wed Feb 06, 2019&gt;</span></span>
-   **Commemorate Version:** First Release.


CHANGELOG
* Version 0.3.0

+ Data :: <2019-03-10 Sun>
+ Add ::
    - Custom ignore when flatten iterable.

* Version 0.2.0

+ Data :: <2019-02-15 Fri>
+ Add ::
    - flatten in extra.

* Version 0.1.0

+ Data :: <2019-02-06 Wed>
+ Commemorate Version :: First Release.


tests/test_nalude.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
r"""
Life's pathetic, have fun ("▔□▔)/hi~♡ Nasy.

Excited without bugs::

    |             *         *
    |                  .                .
    |           .
    |     *                      ,
    |                   .
    |
    |                               *
    |          |\___/|
    |          )    -(             .              ·
    |         =\ -   /=
    |           )===(       *
    |          /   - \
    |          |-    |
    |         /   -   \     0.|.0
    |  NASY___\__( (__/_____(\=/)__+1s____________
    |  ______|____) )______|______|______|______|_
    |  ___|______( (____|______|______|______|____
    |  ______|____\_|______|______|______|______|_
    |  ___|______|______|______|______|______|____
    |  ______|______|______|______|______|______|_
    |  ___|______|______|______|______|______|____

author   : Nasy https://nasy.moe
date     : Jan 30, 2019
email    : Nasy <nasyxx+nalude@gmail.com>
filename : test_nalude.py
project  : nalude
license  : LGPL-3.0+

There are more things in heaven and earth, Horatio, than are dreamt.
 --  From "Hamlet"
"""
# Standard Library
import unittest
from typing import Tuple

# Prelude
from nalude import (o, fst, id_, snd, all_, any_, drop, flip, head, init,
                    last, not_, null, span, tail, take, const, curry, cycle,
                    foldl, foldr, lines, until, unzip, words, break_, concat,
                    foldl1, foldr1, repeat, flatten, iterate, product,
                    splitat, uncurry, unlines, unwords, zipwith, concatmap,
                    dropwhile, replicate, takewhile, __version__,)


class NaludeTest(unittest.TestCase):
    """Nasy prelude package test."""

    def test_version(self) -> None:
        """Test version of nalude."""
        with open("pyproject.toml") as f:
            for line in f:
                if "version" in line:
                    version = line.split()[-1].replace('"', "")
                    break
        self.assertEqual(__version__, version)

    def test_folds_folds(self) -> None:
        """Test folds and traversals."""
        self.assertEqual(
            foldl(lambda x, y: x + y, 0, range(10)),
            foldr(lambda x, y: x + y, 0, range(10)),
        )
        self.assertEqual(
            foldl1(lambda x, y: x + y, range(10)),
            foldr1(lambda x, y: x + y, range(10)),
        )
        self.assertEqual(
            foldl1(lambda x, y: x + y, range(10)),
            foldr(lambda x, y: x + y, 0, range(10)),
        )

    def test_folds_product(self) -> None:
        """Test folds product."""
        self.assertEqual(product([1, 2, 3, 4, 5]), 5 * 4 * 3 * 2 * 1)
        self.assertEqual(product(range(10)), 0)

    def test_lists_head_last(self) -> None:
        """Test lists head and last."""
        self.assertEqual(head(range(1)), last(range(1)))
        self.assertEqual(head("yoooo!"), last("!ooooy"))
        self.assertRaises(IndexError, head, list())
        self.assertRaises(IndexError, last, list())

    def test_lists_tail_init(self) -> None:
        """Test lists tail and init."""
        self.assertEqual(list(tail(range(5))), list(init([1, 2, 3, 4, 0])))
        self.assertEqual(list(tail(range(1))), list(init(range(1))))
        self.assertRaises(IndexError, lambda x: next(tail(x)), list())
        self.assertRaises(IndexError, lambda x: next(init(x)), list())

    def test_lists_null(self) -> None:
        """Test lists null."""
        self.assertTrue(null(range(0)))
        self.assertFalse(null(range(1)))

    def test_lists_iterate(self) -> None:
        """Test lists iterate."""
        for i, e in zip(range(1, 6), iterate(lambda x: x + 1, 1)):
            with self.subTest(i=i, e=e):
                self.assertEqual(e, i)

        for i, e in zip(range(1, 6), iterate(lambda x: [1] + x, [1])):
            with self.subTest(i=i, e=e):
                self.assertEqual(e, [1] * i)

    def test_lists_repeat_replicate(self) -> None:
        """Test lists repeat and replicate."""
        for r1, r2 in zip(replicate(10, 5), repeat(5)):
            with self.subTest(r1=r1, r2=r2):
                self.assertEqual(r1, r2)

    def test_lists_cycle(self) -> None:
        """Test lists cycle."""
        for c1, c2 in zip([1, 2, 3] * 3, cycle([1, 2, 3])):
            with self.subTest(str(c1), c1=c1, c2=c2):
                self.assertEqual(c1, c2)
        self.assertRaises(IndexError, lambda x: next(cycle(x)), list())

    def test_lists_take_drop_splitat(self) -> None:
        """Test lists take, drop and splitat."""
        self.assertEqual(list(take(5, [])), list(drop(5, [])))
        self.assertEqual(list(take(-1, [1, 2, 3])), [])
        self.assertEqual(list(take(1, [1, 2, 3])), [1])
        self.assertEqual(list(drop(-1, [1, 2, 3])), [1, 2, 3])
        self.assertEqual(list(drop(1, [1, 2, 3])), [2, 3])

        self.assertEqual(
            list(take(5, range(5))), list(splitat(5, range(5))[0])
        )
        self.assertEqual(
            list(drop(5, range(5))), list(splitat(5, range(5))[1])
        )
        xs = range(10)
        self.assertEqual(
            (list(take(5, (i for i in xs))), list(drop(5, (i for i in xs)))),
            tuple(map(list, splitat(5, (i for i in xs)))),
        )

    def test_lists_takewhile_dropwhile_span_break(self) -> None:
        """Test lists takewhile and dropwhile, span and break."""
        self.assertEqual(
            list(takewhile(lambda x: x < 3, range(10))), [0, 1, 2]
        )
        self.assertEqual(list(dropwhile(lambda x: x < 8, range(10))), [8, 9])
        self.assertEqual(
            span(lambda x: x < 3, [1, 2, 3, 4, 1, 2, 3, 4]),
            break_(lambda x: x > 2, [1, 2, 3, 4, 1, 2, 3, 4]),
        )
        self.assertEqual(
            span(lambda x: x < 9, [1, 2, 3]),
            break_(lambda x: x > 9, [1, 2, 3]),
        )
        self.assertEqual(
            span(lambda x: x > 9, [1, 2, 3]),
            break_(lambda x: x < 9, [1, 2, 3]),
        )

    def test_miscellaneous_id_const(self) -> None:
        """Test miscellaneous id and const."""
        self.assertEqual(id_(0), 0)
        self.assertEqual(const(0, 1), 0)

    def test_miscellaneous_o_flip(self) -> None:
        """Test miscellaneous o and flip."""

        def f1(x: int) -> int:
            """Function f1."""
            return x * 2

        def f2(x: int) -> int:
            """Function f2."""
            return x + 2

        def f(x: int, y: int) -> str:
            """Function f."""
            return str(x) + str(y)

        self.assertEqual(o(f1, f2)(1), f1(f2(1)))
        self.assertEqual(o(f1, o(f2, f1))(1), f1(f2(f1(1))))
        self.assertEqual(flip(f)(1, 2), f(2, 1))

    def test_miscellaneous_until(self) -> None:
        """Test miscellaneous until."""
        self.assertEqual(until(lambda x: x > 10, lambda x: x * 2, 1), 16)

    def test_strings_lines_unlines(self) -> None:
        """Test strings lines and unlines."""
        self.assertEqual(lines(""), [])
        self.assertEqual(lines("\n"), [""])
        self.assertEqual(lines("one\n"), ["one"])
        self.assertEqual(lines("one\n\n"), ["one", ""])
        self.assertEqual(lines("one\ntwo"), ["one", "two"])
        self.assertEqual(lines("one\ntwo\n"), ["one", "two"])
        self.assertEqual(unlines(lines("one\ntwo")), "one\ntwo\n")

    def test_strings_words_unwords(self) -> None:
        """Test strings words and unwords."""
        self.assertEqual(
            words("Lorem ipsum\ndolor"), ["Lorem", "ipsum", "dolor"]
        )
        self.assertEqual(
            unwords(["Lorem", "ipsum", "dolor"]), "Lorem ipsum dolor"
        )

    def test_specials_not(self) -> None:
        """Test specials not."""
        self.assertTrue(not_(bool)(False))
        self.assertTrue(not_(bool)([]))
        self.assertFalse(not_(bool)(True))

    def test_specials_any_all(self) -> None:
        """Test specials any and all."""
        self.assertTrue(any_(lambda x: x >= 5, range(6)))
        self.assertFalse(all_(lambda x: x < 5, range(6)))

    def test_special_concat_map(self) -> None:
        """Test specials concat and concatmap."""
        self.assertEqual(
            list(concat([[1, 2, 3], [2, 3, 4]])), [1, 2, 3, 2, 3, 4]
        )
        self.assertEqual(
            list(concat(((i for i in ii) for ii in [[1, 2, 3], [2, 3, 4]]))),
            [1, 2, 3, 2, 3, 4],
        )
        self.assertEqual(
            list(
                concatmap(
                    lambda x: [1] + list(x), [[1, 2, 3, 4], [2, 3, 4, 5]]
                )
            ),
            [1, 1, 2, 3, 4, 1, 2, 3, 4, 5],
        )

    def test_tuple_fst_snd(self) -> None:
        """Test tuple fst and snd."""
        t = (1, 2)
        self.assertEqual((fst(t), snd(t)), t)

    def test_tuple_curry_uncurry(self) -> None:
        """Test tuple curry and uncurry."""

        def f(ab: Tuple[int, str]) -> Tuple[str, int]:
            """Function f."""
            return str(ab[0]), int(ab[1])

        def f_(a: str, b: int) -> Tuple[int, str]:
            """Function f."""
            return int(a), str(b)

        self.assertEqual(curry(f, 1, "2"), ("1", 2))
        self.assertEqual(uncurry(f_, ("1", 2)), (1, "2"))

    def test_zip_and_unzip(self) -> None:
        """Test zipwith and unzip."""

        def fzw(a: int, b: int) -> int:
            """Function zipwith."""
            return a + b

        self.assertEqual(
            list(zipwith(fzw, range(10), range(2, 12))),
            list(map(lambda xy: xy[0] + xy[1], zip(range(10), range(2, 12)))),
        )
        self.assertEqual(
            unzip([(1, 2), (2, 3), (3, 4)]), ([1, 2, 3], [2, 3, 4])
        )

    def test_extra_flatten(self) -> None:
        """Test extra flatten."""
        self.assertEqual(
            list(
                flatten(
                    [
                        1,
                        2,
                        3,
                        4,
                        "123",
                        [
                            1,
                            2,
                            3,
                            [
                                dict(a=12),
                                ["12", "12", (1, 2, 3), (i for i in range(3))],
                            ],
                        ],
                    ],
                    ignore=tuple(),
                )
            ),
            [1, 2, 3, 4, "123", 1, 2, 3, "a", "12", "12", 1, 2, 3, 0, 1, 2],
        )
        self.assertEqual(
            list(
                flatten(
                    [
                        1,
                        2,
                        3,
                        4,
                        "123",
                        [
                            1,
                            2,
                            3,
                            [
                                dict(a=12),
                                ["12", "12", (1, 2, 3), (i for i in range(3))],
                            ],
                        ],
                    ],
                    ignore=(dict, tuple),
                )
            ),
            [
                1,
                2,
                3,
                4,
                "123",
                1,
                2,
                3,
                {"a": 12},
                "12",
                "12",
                (1, 2, 3),
                0,
                1,
                2,
            ],
        )
        self.assertEqual(
            list(
                flatten(
                    [
                        1,
                        2,
                        3,
                        4,
                        "123",
                        [
                            1,
                            2,
                            3,
                            [
                                dict(a=12),
                                ["12", "12", (1, 2, 3), (i for i in range(3))],
                            ],
                        ],
                    ]
                )
            ),
            [
                1,
                2,
                3,
                4,
                "123",
                1,
                2,
                3,
                dict(a=12),
                "12",
                "12",
                1,
                2,
                3,
                0,
                1,
                2,
            ],
        )


def run() -> None:
    """Run test."""
    unittest.main()


if __name__ == "__main__":
    run()


nalude/extra.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
r"""
Life's pathetic, have fun ("▔□▔)/hi~♡ Nasy.

Excited without bugs::

    |             *         *
    |                  .                .
    |           .
    |     *                      ,
    |                   .
    |
    |                               *
    |          |\___/|
    |          )    -(             .              ·
    |         =\ -   /=
    |           )===(       *
    |          /   - \
    |          |-    |
    |         /   -   \     0.|.0
    |  NASY___\__( (__/_____(\=/)__+1s____________
    |  ______|____) )______|______|______|______|_
    |  ___|______( (____|______|______|______|____
    |  ______|____\_|______|______|______|______|_
    |  ___|______|______|______|______|______|____
    |  ______|______|______|______|______|______|_
    |  ___|______|______|______|______|______|____

author   : Nasy https://nasy.moe
date     : Feb 15, 2019
email    : Nasy <nasyxx+nalude@gmail.com>
filename : extra.py
project  : nalude
license  : LGPL-3.0+

There are more things in heaven and earth, Horatio, than are dreamt.
 --  From "Hamlet"
"""
# Standard Library
from typing import Type, Tuple, Iterable

# Local Packages
from .types import G, a


def flatten(xs: a, *, ignore: Tuple[Type, ...] = (dict,)) -> G:
    """Flatten iterables of iterable to a single one.

    The true type of it is:: `T = Iterable[Untion[a, T]]` ."""
    if not (isinstance(xs, (str, *ignore)) or not isinstance(xs, Iterable)):
        for x in xs:
            for x_ in flatten(x, ignore=ignore):
                yield x_
    else:
        yield xs


nalude/types.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
r"""
Life's pathetic, have fun ("▔□▔)/hi~♡ Nasy.

Excited without bugs::

    |             *         *
    |                  .                .
    |           .
    |     *                      ,
    |                   .
    |
    |                               *
    |          |\___/|
    |          )    -(             .              ·
    |         =\ -   /=
    |           )===(       *
    |          /   - \
    |          |-    |
    |         /   -   \     0.|.0
    |  NASY___\__( (__/_____(\=/)__+1s____________
    |  ______|____) )______|______|______|______|_
    |  ___|______( (____|______|______|______|____
    |  ______|____\_|______|______|______|______|_
    |  ___|______|______|______|______|______|____
    |  ______|______|______|______|______|______|_
    |  ___|______|______|______|______|______|____

author   : Nasy https://nasy.moe
date     : Jan 31, 2019
email    : Nasy <nasyxx+nalude@gmail.com>
filename : types.py
project  : nalude
license  : LGPL-3.0+

There are more things in heaven and earth, Horatio, than are dreamt.
 --  From "Hamlet"
"""
# Standard Library
from typing import TypeVar, Callable, Generator

a = TypeVar("a")
b = TypeVar("b")
c = TypeVar("c")
t = TypeVar("t")
G = Generator[a, None, None]
Num = TypeVar("Num", int, float, complex)
Predicate = Callable[[a], bool]

__all__ = ["a", "b", "c", "t", "G", "Num", "Predicate"]


nalude/__init__.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
r"""
Life's pathetic, have fun ("▔□▔)/hi~♡ Nasy.

Excited without bugs::

    |             *         *
    |                  .                .
    |           .
    |     *                      ,
    |                   .
    |
    |                               *
    |          |\___/|
    |          )    -(             .              ·
    |         =\ -   /=
    |           )===(       *
    |          /   - \
    |          |-    |
    |         /   -   \     0.|.0
    |  NASY___\__( (__/_____(\=/)__+1s____________
    |  ______|____) )______|______|______|______|_
    |  ___|______( (____|______|______|______|____
    |  ______|____\_|______|______|______|______|_
    |  ___|______|______|______|______|______|____
    |  ______|______|______|______|______|______|_
    |  ___|______|______|______|______|______|____

author   : Nasy https://nasy.moe
date     : Jan 30, 2019
email    : Nasy <nasyxx+nalude@gmail.com>
filename : __init__.py
project  : nalude
license  : LGPL-3.0+

There are more things in heaven and earth, Horatio, than are dreamt.
 --  From "Hamlet"
"""
__version__ = "0.3.0"
# Local Packages
from .extra import flatten
from .nalude import (o, fst, id_, snd, all_, any_, drop, flip, head, init,
                     last, not_, null, span, tail, take, const, curry,
                     cycle, foldl, foldr, lines, until, unzip, words,
                     break_, concat, foldl1, foldr1, repeat, iterate,
                     product, splitat, uncurry, unlines, unwords, zipwith,
                     concatmap, dropwhile, replicate, takewhile,)

if __name__ == "__main__":
    # Folds and Traversals
    assert foldl
    assert foldl1
    assert foldr
    assert foldr1
    assert product

    # Lists
    assert head
    assert last
    assert init
    assert tail
    assert null
    assert iterate
    assert repeat
    assert replicate
    assert cycle
    assert take
    assert drop
    assert splitat
    assert takewhile
    assert dropwhile
    assert span
    assert break_

    # Miscellaneous
    assert id_
    assert const
    assert flip
    assert o
    assert until

    # String
    assert lines
    assert unlines
    assert words
    assert unwords

    # Specials
    assert not_
    assert all_
    assert any_
    assert concat
    assert concatmap

    # TUple
    assert fst
    assert snd
    assert curry
    assert uncurry

    # Zip and Unzip
    assert zipwith
    assert unzip

    # Extra
    assert flatten


nalude/nalude.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
r"""
Life's pathetic, have fun ("▔□▔)/hi~♡ Nasy.

Excited without bugs::

    |             *         *
    |                  .                .
    |           .
    |     *                      ,
    |                   .
    |
    |                               *
    |          |\___/|
    |          )    -(             .              ·
    |         =\ -   /=
    |           )===(       *
    |          /   - \
    |          |-    |
    |         /   -   \     0.|.0
    |  NASY___\__( (__/_____(\=/)__+1s____________
    |  ______|____) )______|______|______|______|_
    |  ___|______( (____|______|______|______|____
    |  ______|____\_|______|______|______|______|_
    |  ___|______|______|______|______|______|____
    |  ______|______|______|______|______|______|_
    |  ___|______|______|______|______|______|____

author   : Nasy https://nasy.moe
date     : Jan 31, 2019
email    : Nasy <nasyxx+nalude@gmail.com>
filename : nalude.py
project  : nalude
license  : LGPL-3.0+

There are more things in heaven and earth, Horatio, than are dreamt.
 --  From "Hamlet"
"""
# Standard Library
from typing import List, Tuple, Callable, Iterable, Generator
from functools import reduce

# Local Packages
from .types import G, Num, Predicate, a, b, c

# ----------------------------------------------------------------------
# Folds and Traversals
# ----------------------------------------------------------------------


def foldl(f: Callable[[b, a], b], init: b, t: Iterable[a]) -> b:
    """Left-associative fold of an iterable."""
    return reduce(f, t, init)


def foldr(f: Callable[[a, b], b], init: b, t: Iterable[a]) -> b:
    """Right-associative fold of an iterable."""
    t_ = list(t)
    return reduce(flip(f), reversed(t_), init)


def foldl1(f: Callable[[a, a], a], t: Iterable[a]) -> a:
    """Left-associative fold of an iterable.

    This is avariant of foldl without base case.
    """
    t_ = iter(t)
    try:
        return reduce(f, t_, next(t_))
    except StopIteration:
        raise IndexError("Iterable is Empty")


def foldr1(f: Callable[[a, a], a], t: Iterable[a]) -> a:
    """Right-associative fold of an iterable.

    This is a variant of foldr without base case.
    """
    t_ = iter(reversed(list(t)))
    try:
        return reduce(f, t_, next(t_))
    except StopIteration:
        raise IndexError("Iterable is Empty")


def product(nums: Iterable[Num]) -> Num:
    """Computes the product of the numbers."""
    try:
        return foldl1(lambda b, a: a * b, nums)
    except IndexError:
        raise IndexError("Numbers are empty")


# ----------------------------------------------------------------------
# Lists
# ----------------------------------------------------------------------


def head(xs: Iterable[a]) -> a:
    """Extract the first element of an iterable."""
    try:
        return next(iter(xs))
    except StopIteration:
        raise IndexError("Iterable is empty.")


def tail(xs: Iterable[a]) -> G:
    """Extract the elements after the head of an iterable."""
    try:
        ixs = iter(xs)
        next(ixs)
        for x in ixs:
            yield x
    except StopIteration:
        raise IndexError("Iterable is empty.")


def last(xs: Iterable[a]) -> a:
    """Extract the last element of an iterable."""
    try:
        return next(iter(reversed(list(xs))))
    except StopIteration:
        raise IndexError("Iterable is empty.")


def init(xs: Iterable[a]) -> G:
    """Extract all the elements of an iterable except the last one."""
    ixs = iter(xs)
    try:
        x_ = next(ixs)
        n = True
        while n:
            try:
                x = next(ixs)
                yield x_
                x_ = x
            except StopIteration:
                n = False
    except StopIteration:
        raise IndexError("Iterable is empty.")


def null(l: Iterable[a]) -> bool:
    """Test whether the sequence is empty."""
    try:
        next(iter(l))
        return False
    except StopIteration:
        return True


def iterate(f: Callable[[a], a], x: a) -> G:
    """Yield a tuple of repeated applications of f to x."""
    while True:
        yield x
        x = f(x)


def repeat(x: a) -> G:
    """Repeat x."""
    while True:
        yield x


def replicate(n: int, x: a) -> G:
    """Return a iterator of length n with x the value of every element."""
    for _ in range(n):
        yield x


def cycle(xs: Iterable[a]) -> G:
    """Tie an iterable to infinite circuler one."""
    xs_ = []  # if x is a finite iterator, this can save the value.
    ixs = iter(xs)
    try:
        x = next(ixs)
        yield x
        xs_.append(x)
    except StopIteration:
        raise IndexError("Iterable is empty.")
    for x in ixs:
        yield x
        xs_.append(x)
    while True:
        for x in xs_:
            yield x


def take(n: int, xs: Iterable[a]) -> G:
    """Return the first n elements of sequence xs."""
    for _, x in zip(range(n), xs):
        yield x


def drop(n: int, xs: Iterable[a]) -> G:
    """Return the remaining elements of xs after the first n elements."""
    ixs = iter(xs)
    try:
        for _ in range(n):
            next(ixs)
        while True:
            yield next(ixs)
    except StopIteration:
        pass


def splitat(n: int, xs: Iterable[a]) -> Tuple[G, G]:
    """Equal to (take(n, xs), drop(n, xs)).

    Return a tuple where the first element is xs prefix of length n and second
    element is the remainder of the sequence xs.
    """
    ixs = iter(xs)
    return (i for i in list(take(n, ixs))), (i for i in ixs)


def takewhile(p: Predicate, xs: Iterable[a]) -> G:
    """Return the longest prefix of xs of elements that satisfy predicate p."""
    for x in xs:
        if p(x):
            yield x
        else:
            break


def dropwhile(p: Predicate, xs: Iterable[a]) -> G:
    """Returns the suffix remaining after takewhile(p, xs)."""
    ixs = iter(xs)
    for x in ixs:
        if not p(x):
            yield x
            break
    for x in ixs:
        yield x


def span(p: Predicate, xs: Iterable[a]) -> Tuple[List[a], List[a]]:
    """Equal to (takewhile(p, xs), dropwhile(p, xs)).

    Note: Not lazy."""
    lxs = list(xs)
    return list(takewhile(p, lxs)), list(dropwhile(p, lxs))


def break_(p: Predicate, xs: Iterable[a]) -> Tuple[List[a], List[a]]:
    """Equal to (takewhile(not_(p), xs), dropwhile(not_(p), xs)).

    Note: Not lazy."""
    lxs = list(xs)
    return list(takewhile(not_(p), lxs)), list(dropwhile(not_(p), lxs))


# ----------------------------------------------------------------------
# Miscellaneous
# ----------------------------------------------------------------------


def id_(x: a) -> a:
    """Identity function."""
    return x


def const(x: a, _: b) -> a:
    """Evaluates to x for all inputs."""
    return x


def o(f1: Callable[[b], c], f2: Callable[[a], b]) -> Callable[[a], c]:
    """Function composition.

    >>> o(f1, f2)(a) == f1(f2(a))
    """
    return lambda a_: f1(f2(a_))


def flip(f: Callable[[a, b], c]) -> Callable[[b, a], c]:
    """Flip takes its two arguments into the reverse order of f.

    >>> flip(f)(b, a) == f(a, b)
    """
    return lambda b_, a_: f(a_, b_)


def until(p: Predicate, f: Callable[[a], a], x: a) -> a:
    """Yields the result of applying f until p holds."""
    return p(x) and x or until(p, f, f(x))


# ----------------------------------------------------------------------
# Strings
# ----------------------------------------------------------------------


def lines(s: str) -> List[str]:
    """Break up a string into a list of strings at newline characters."""
    return s.splitlines()


def unlines(xs: Iterable[str]) -> str:
    """The inverse operation of lines, append a newline to each."""
    return "\n".join(xs) + ("\n" if xs else "")


def words(s: str) -> List[str]:
    """Breaks a string up into a list of words, which were delimited by white
    space."""
    return s.split()


def unwords(xs: Iterable[str]) -> str:
    """The inverse operation of words, join words with space."""
    return " ".join(xs)


# ----------------------------------------------------------------------
# Specials
# ----------------------------------------------------------------------


def not_(f: Callable[..., bool]) -> Callable[..., bool]:
    """Boolean "not"."""

    def _wrap(*args: a, **kwds: b) -> bool:
        """Wrapper of not_."""
        return not f(*args, **kwds)

    return _wrap


def all_(p: Predicate, xs: Iterable[a]) -> bool:
    """Determine whether all elements of the structure satisfy the p."""
    return all(map(p, xs))


def any_(p: Predicate, xs: Iterable[a]) -> bool:
    """Determine whether any element of the structure satisfies the p."""
    return any(map(p, xs))


def concat(xss: Iterable[Iterable[a]]) -> G:
    """The concatenation of all the elements of a container of Iterables."""
    for xs in xss:
        for x in xs:
            yield x


def concatmap(
    f: Callable[[Iterable[a]], Iterable[b]], xss: Iterable[Iterable[a]]
) -> Generator[b, None, None]:
    """Map a function over all the elements of a container and concatenate the
    resulting lists."""
    for xs in xss:
        for x in f(xs):
            yield x


# ----------------------------------------------------------------------
# Tuples
# ----------------------------------------------------------------------


def fst(t: Tuple[a, b]) -> a:
    """Extract the first component of a tuple."""
    return t[0]


def snd(t: Tuple[a, b]) -> b:
    """Extract the second component of a tuple."""
    return t[1]


def curry(f: Callable[[Tuple[a, b]], c], a_: a, b_: b) -> c:
    """Converts an uncurried function to a curried function."""
    return f((a_, b_))


def uncurry(f: Callable[[a, b], c], ab: Tuple[a, b]) -> c:
    """Converts a curried function to a function on pairs."""
    a_, b_ = ab
    return f(a_, b_)


# ----------------------------------------------------------------------
# Zip and Unzip
# ----------------------------------------------------------------------


def zipwith(
    f: Callable[..., c], *seqs: Iterable[a]
) -> Generator[c, None, None]:
    """Zipwith is map(f, zip), but f accept separate args instead of tuple"""
    for zseq in zip(*seqs):
        yield f(*zseq)


def unzip(pairs: Iterable[Tuple[a, ...]]) -> Tuple[Iterable[a], ...]:
    """Transform an iterable of pairs into a tuple of sequence.

    Note: Not lazy."""
    pairs = list(pairs)
    if pairs:
        return tuple(
            [pair[i] for pair in pairs] for i in range(len(head(pairs)))
        )
    return tuple()
